### PR TITLE
Adds app commands for devfile components

### LINKED ISSUE #3402: `odo app` support for devfile components
/kind user-story

## User Story
As a user, I want to group different devfile components to an `app` similar to what I can do for s2i components.

## Background
odo has the `odo app` command for grouping existing s2i components under an app. 

## Acceptance Criteria 
- [x] It should behave the same as the existing `odo app` command but allowing devfile components to be added to the app
- [x] It should support `--app appname` parameter on `odo create` command for devfile components

We need to figure out where we store the app-component relationship data.

/kind user-story
/area devfile

### PR DESCRIPTION
**Pull request by mik-dass**
**What type of PR is this?**

/kind feature

**What does does this PR do / why we need it**:

This PR adds app support for devfile components,

**Which issue(s) this PR fixes**:

Fixes #3402 

**PR acceptance criteria**:

- [ ] Unit test 

- [x] Integration test 

- [ ] Documentation 

- [X] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:

Create some components (s2i and devfile) in different apps and check

- `odo app list`
- `odo app describe <app-name>`
- `odo app delete <app-name>`

### CODE REVIEW

File: pkg/application/application.go [hunk 1]
```diff
@@ -33,13 +34,27 @@ func ListInProject(client *occlient.Client) ([]string, error) {
...
+	deploymentSupported, err := client.IsDeploymentConfigSupported()
+	if err != nil {
+		return nil, err
+	}
+
 	// Get all DeploymentConfigs with the "app" label
-	deploymentConfigAppNames, err := client.GetDeploymentConfigLabelValues(applabels.ApplicationLabel, applabels.ApplicationLabel)
+	deploymentAppNames, err := client.GetKubeClient().GetDeploymentLabelValues(applabels.ApplicationLabel, applabels.ApplicationLabel)
 	if err != nil {
 		return nil, errors.Wrap(err, "unable to list applications from deployment config")
```
Comment (adisky) [inline pkg/application/application.go:45]: s/unable to list applications from deployment config/unable to list applications from deployment

File: tests/integration/devfile/cmd_devfile_app_test.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+		err := os.Setenv("KUBECONFIG", originalKubeconfig)
+		Expect(err).NotTo(HaveOccurred())
+		helper.DeleteDir(context)
+		os.Unsetenv("GLOBALODOCONFIG")
+	})
+
+	Context("when running help for app command", func() {
+		It("should display the help", func() {
+			appHelp := helper.CmdShouldPass("odo", "app", "-h")
+			Expect(appHelp).To(ContainSubstring("Performs application operations related to your project."))
```
Comment (adisky) [inline tests/integration/devfile/cmd_devfile_app_test.go:43]: do you think we should add test for help? we always consciously change help messages.

File: pkg/kclient/kclient.go [hunk 3]
```diff
@@ -92,3 +107,76 @@ func CreateObjectMeta(name, namespace string, labels, annotations map[string]str
...
+	if err != nil {
+		return err
+	}
+	defer watcher.Stop()
+	eventCh := watcher.ResultChan()
+
+	for {
+		select {
+		case event, ok := <-eventCh:
+			_, typeOk := event.Object.(*appsv1.DeploymentConfig)
```
Comment (adisky) [inline pkg/kclient/kclient.go:166]: DeploymentConfig in Kclient?

**Review by adisky — COMMENTED** [review]
Works for me, tried `odo create nodejs --app`, `odo app list`, `odo app describe`  and `odo app delete ` commands

File: tests/integration/devfile/cmd_devfile_app_test.go (hunk 2 above)
Comment (mik-dass) [reply tests/integration/devfile/cmd_devfile_app_test.go:43]: It was there in the s2i version of the test script, so I kept it.

**Comment by mik-dass** [comment]
@adisky Fixed

File: tests/integration/devfile/cmd_devfile_app_test.go (hunk 2 above)
Comment (adisky) [reply tests/integration/devfile/cmd_devfile_app_test.go:43]: I think there is no need, what you say?

Comment (mik-dass) [reply tests/integration/devfile/cmd_devfile_app_test.go]: removed as the test in the s2i version is sufficient.

**Comment by mik-dass** [comment]
```
error: build error: failed to pull image: After retrying
```
/retest

File: pkg/kclient/deployments.go [hunk 4]
```diff
@@ -3,6 +3,7 @@ package kclient
 import (
 	"encoding/json"
 	"fmt"
+	sort "sort"
```
Comment (adisky) [inline pkg/kclient/deployments.go:6]: alias?

File: pkg/kclient/kclient.go [hunk 5]
```diff
@@ -92,3 +107,76 @@ func CreateObjectMeta(name, namespace string, labels, annotations map[string]str
...
+	}
+	// Delete Deployments
+	klog.V(3).Info("Deleting Deployments")
+	err := c.appsClient.Deployments(c.Namespace).DeleteCollection(&metav1.DeleteOptions{PropagationPolicy: &deletionPolicy}, metav1.ListOptions{LabelSelector: selector})
+	if err != nil {
+		errorList = append(errorList, "unable to delete deployments")
+	}
+
+	// for --wait it waits for component to be deleted
+	// TODO: Need to modify for `odo app delete`, currently wait flag is added only in `odo component delete`
```
Comment (adisky) [inline pkg/kclient/kclient.go:133]: is there something remaining  `odo app delete`?

**Review by adisky — COMMENTED** [review]
/lgtm

File: pkg/kclient/deployments.go (hunk 4 above)
Comment (mik-dass) [reply pkg/kclient/deployments.go]: Fixed

File: pkg/kclient/kclient.go (hunk 5 above)
Comment (mik-dass) [reply pkg/kclient/kclient.go:133]: Most of the function is copied from occlient.go. This `TODO` is also present there so I kept it.

**Pull request merged**